File: Android/scenes/ktv_singbattle/src/main/java/io/agora/scene/ktv/singbattle/live/listener/LrcActionListenerImpl.java
package io.agora.scene.ktv.singbattle.live.listener;

import android.content.Context;

import io.agora.karaoke_view.v11.model.LyricsLineModel;
import io.agora.karaoke_view.v11.model.LyricsModel;
import io.agora.scene.ktv.singbattle.live.RoomLivingViewModel;
import io.agora.scene.ktv.singbattle.widget.lrcView.LrcControlView;

public class LrcActionListenerImpl implements LrcControlView.OnKaraokeEventListener {

    private final Context mContext;
    private final RoomLivingViewModel mViewModel;
    private final LrcControlView mLrcControlView;

    public LrcActionListenerImpl(Context context, RoomLivingViewModel viewModel, LrcControlView lrcControlView) {
        mContext = context;
        mViewModel = viewModel;
        mLrcControlView = lrcControlView;
    }


    @Override
    public void onSwitchOriginalClick() {
        LrcControlView.OnKaraokeEventListener.super.onSwitchOriginalClick();
        mViewModel.musicToggleOriginal();

        mLrcControlView.setSwitchOriginalChecked(mViewModel.isOriginalMode());
    }

    @Override
    public void onPlayClick() {
        LrcControlView.OnKaraokeEventListener.super.onPlayClick();

        mViewModel.musicToggleStart();
    }

    @Override
    public void onDragTo(long time) {
        mViewModel.musicSeek(time);
    }

    @Override
    public void onLineFinished(LyricsLineModel line, int score, int cumulativeScore, int index, int total) {
        mLrcControlView.updateScore(score, cumulativeScore, /** Workaround(Hai_Guo)*/total * 100);
        mViewModel.syncSingleLineScore(score, cumulativeScore, index, total * 100);
    }

    @Override
    public void onReGetLrcUrl() {
        mViewModel.reGetLrcUrl();
    }

    @Override
    public void onGraspSongClick() {
        mViewModel.graspSong();
    }
}

File: Android/scenes/cantata/src/main/java/io/agora/scene/cantata/live/listener/LrcActionListenerImpl.kt
package io.agora.scene.cantata.live.listener

import android.content.Context
import io.agora.karaoke_view_ex.internal.model.LyricsLineModel
import io.agora.scene.cantata.live.RoomLivingActivity
import io.agora.scene.cantata.live.RoomLivingViewModel
import io.agora.scene.cantata.widget.lrcView.LrcControlView

/**
 * Mixer Control
 */
open class LrcActionListenerImpl constructor(
    private val mContext: Context,
    private val mViewModel: RoomLivingViewModel,
    private val mLrcControlView: LrcControlView
) : LrcControlView.OnKaraokeEventListener {
    override fun onSwitchOriginalClick() {
        super.onSwitchOriginalClick()
        mViewModel.musicToggleOriginal()
        mLrcControlView.setSwitchOriginalChecked(mViewModel.isOriginalMode())
    }

    override fun onPlayClick() {
        mViewModel.musicToggleStart()
    }

    override fun onStartSing() {
        mViewModel.leaveChorus()
    }

    override fun onJoinChorus() {
        if (mContext is RoomLivingActivity) {
            mContext.toggleSelfAudio(true, callback = {
                mViewModel.joinChorus()
            })
        }
    }

    override fun onLeaveChorus() {
        mViewModel.leaveChorus()
    }

    override fun onDragTo(time: Long) {
        mViewModel.musicSeek(time)
    }

    override fun onLineFinished(line: LyricsLineModel?, score: Int, cumulativeScore: Int, index: Int, total: Int) {
        mViewModel.updateSeatScoreStatus(score, cumulativeScore)
    }

    override fun onSkipPreludeClick() {
        val lyrics = mLrcControlView.karaokeView?.lyricData ?: return
        // Experience will be better when seeking 2000 milliseconds ahead
        val seekPosition = lyrics.preludeEndPosition - 2000
        mViewModel.musicSeek(if (seekPosition > 0) seekPosition else 0)
    }

    override fun onSkipPostludeClick() {
        val lyrics = mLrcControlView.karaokeView?.lyricData ?: return
        mViewModel.musicSeek(mViewModel.getSongDuration()!! - 500)
    }

    override fun onReGetLrcUrl() {
        mViewModel.reGetLrcUrl()
    }
}